src/pools/ID3MPool.sol

```diff
@@ -0,0 +1,120 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+/**
+    @title D3M Pool Interface
+    @notice Pool contracts are contracts that the Hub uses to standardize
+    interactions with external Pools.
+    @dev Implementing contracts will hold any balance provided by the external
+    pool as well as the balance in the Vat. This interface aims to use EIP-4626
+    guidelines for assets/shares/maxWithdraw etc.
+*/
+interface ID3MPool {
+    /**
+        @notice Hopes on an address in the Vat.
+        @param hub address you want to hope on
+    */
+    function hope(address hub) external;
+
+    /**
+        @notice Nopes on an address in the Vat.
+        @param hub address you want to nope on
+    */
+    function nope(address hub) external;
+
+    /**
+        @notice Deposit assets (Dai) in the external pool.
+        @dev If the external pool requires a different amount to be passed in, the
+        conversion should occur here as the Hub passes Dai [wad] amounts.
+        msg.sender must be authorized.
+        @param wad amount in asset (Dai) terms that we want to deposit
+        @return bool whether the withdraw was successful
+    */
+    function deposit(uint256 wad) external returns (bool);
+
+    /**
+        @notice Withdraw assets (Dai) from the external pool.
+        @dev If the external pool requires a different amount to be passed in
+        the conversion should occur here as the Hub passes Dai [wad] amounts.
+        msg.sender must be authorized.
+        @param wad amount in asset (Dai) terms that we want to withdraw
+        @return bool whether the withdraw was successful
+    */
+    function withdraw(uint256 wad) external returns (bool);
+
+     /**
+        @notice Transfer shares.
+        @dev If the external pool/shares contract requires a different amount to be
+        passed in the conversion should occur here as the Hub passes Gem [wad]
+        amounts. msg.sender must be authorized.
+        @param dst address that should receive the shares
+        @param wad amount in Gem terms that we want to withdraw
+        @return bool whether the transfer was successful per ERC-20 standard
+    */
+    function transfer(address dst, uint256 wad) external returns (bool);
+
+    /**
+        @notice Transfer all shares from this pool.
+        @dev msg.sender must be authorized.
+        @param dst address that should receive the shares.
+        @return bool whether the transfer was successful per ERC-20 standard
+    */
+    function transferAll(address dst) external returns (bool);
+
+    /// @notice Some external pools require actions before debt changes
+    function preDebtChange() external;
```

Comment:
This is also getting called in `reap`, where it may necessitate a different course of action.

We should give this function a `bytes32` arg `what` and in the hub, pass it something like `prehook("exec");` and `prehook("reap");`. That way a pool implementation can either add conditional logic based on where it's being called or just ignore the param if the hook is the same for both invocations.

Reply:
done ae57751